movieApp/movieApp/Scenes/Main/MainViewModel.swift

```diff
@@ -9,20 +9,113 @@ import UIKit
+    var genreUserDefault = GenreUserDefault()
+
+
+    init() {
+        getMovieGenres()
+    }
+
+    func getPopularMovies(completion: @escaping([CardViewWithImageAndDetailsUIModel]) -> Void) {
+        var cardViewArray: [CardViewWithImageAndDetailsUIModel] = []
+        networking.performRequest(url: APIConstants.popularMovieURL) {(result: Result<PopularMoviesResponseModel, Error>) in
+            switch result {
+            case .success(let popularMoviesResponseModel):
+                for popularMovies in popularMoviesResponseModel.results {
+                    let titleText = popularMovies.title
+                    let voteAverage = String(format: "%.1f", popularMovies.voteAverage)
+                    let releaseDate = popularMovies.releaseDate
+                    let genreIds = popularMovies.genreIds
+                    let detailsText = self.genresIdToString(intArray: genreIds)
+
+                    let cardViewModel = CardViewWithImageAndDetailsUIModel(
+                        coverImage: .dummyImage,
+                        titleText: titleText,
+                        detailsText: detailsText,
+                        imageWithLabelView: ImageWithLabelUIModel(
+                            image: .calendarIcon,
+                            labelText: releaseDate
+                        ),
+                        iconWithPointLabel: IconWithPointLabelUIModel(
+                            icon: .starIcon,
+                            pointNumberText: voteAverage
+                        )
+                    )
+                    cardViewArray.append(cardViewModel)
+                }
+                completion(cardViewArray)
+            case .failure(let error):
+                print(error)
+            }
+        }
+    }
+
+    func getMovieGenres() {
+        if genreUserDefault.checkUserDefault(key: UserDefaultsConstant.genresDict) {
+            return
+        }
+        networking.performRequest(url: APIConstants.movieGenresURL) { (result: Result<MovieGenresResponseModel, Error>) in
+            var movieGenresDict: [String: String] = [:]
+            switch result {
+            case .success(let genresResponseModel):
+                for movieGenres in genresResponseModel.genres {
+                    let id = movieGenres.id.description
+                    let name = movieGenres.name
+                    movieGenresDict[id] = name
+                }
+                genreUserDefault.setUserDefault(value: movieGenresDict, key: UserDefaultsConstant.genresDict)
+            case .failure(let error):
+                print(error)
+            }
+        }
+    }
+
+    func genresIdToString(intArray: [Int]) -> String {
+        var detailsString = ""
+        var editedString = ""
+        if genreUserDefault.checkUserDefault(key: UserDefaultsConstant.genresDict) {
+            let dict = genreUserDefault.getUserDefault(key: UserDefaultsConstant.genresDict)
+            for index in 0..<(intArray.count) {
```

Comment:
I remember that @gidercagri  said "you should check and learn for-each loop and change this old index-based  for loop to for-each loop" so please check it and change this loop

Reply:
fixed.